rllib/utils/exploration/random_encoder.py

```diff
@@ -0,0 +1,338 @@
+        step (int): Current training iteration.
+
+    Returns:
+        Updated beta as per input schedule.
+    """
+    if beta_schedule == "linear_decay":
+        return beta * ((1.0 - rho)**step)
+    return beta
+
+
+def compute_states_entropy(obs_embeds: np.ndarray, embed_dim: int,
+                           k_nn: int) -> np.ndarray:
+    """Compute states entropy using K nearest neighbour method.
+
+    Args:
+        obs_embeds (np.ndarray): Observation latent representation using
+            encoder model.
+        embed_dim (int): Embedding vector dimension.
+        k_nn (int): Number of nearest neighbour for K-NN estimation.
+
+    Returns:
+        Computed states entropy.
+    """
+    obs_embeds_ = np.reshape(obs_embeds, [-1, embed_dim])
+    dist = np.linalg.norm(
+        obs_embeds_[:, None, :] - obs_embeds_[None, :, :], axis=-1)
+    return dist.argsort(axis=-1)[:, :k_nn][:, -1]
+
+
+class RE3UpdateCallbacks():
+    """Update input callbacks to mutate batch with states entropy rewards."""
+
+    _step = 0
+
+    def __init__(self,
+                 *args,
+                 embeds_dim: int = 128,
+                 k_nn: int = 50,
+                 beta: float = 0.1,
+                 rho: float = 0.0001,
+                 beta_schedule: str = "constant",
+                 **kwargs):
+        self.embeds_dim = embeds_dim
+        self.k_nn = k_nn
+        self.beta = beta
+        self.rho = rho
+        self.beta_schedule = beta_schedule
+        self._rms = MovingMeanStd()
+        super().__init__(*args, **kwargs)
+
+    def on_learn_on_batch(
+            self,
+            *,
+            policy: "Policy",
+            train_batch: SampleBatch,
+            result: dict,
+            **kwargs,
+    ):
+        # import Postprocessing here to avoid circular import issue
+        from ray.rllib.evaluation.postprocessing import Postprocessing
+
+        states_entropy = compute_states_entropy(
+            train_batch[SampleBatch.OBS_EMBEDS], self.embeds_dim, self.k_nn)
+        states_entropy = update_beta(
+            self.beta_schedule, self.beta, self.rho,
+            RE3UpdateCallbacks._step) * np.reshape(
+                self._rms(states_entropy),
+                train_batch[SampleBatch.OBS_EMBEDS].shape[:-1],
+            )
+        train_batch[SampleBatch.REWARDS] = (
+            train_batch[SampleBatch.REWARDS] + states_entropy)
+        if Postprocessing.ADVANTAGES in train_batch:
+            train_batch[Postprocessing.ADVANTAGES] = (
+                train_batch[Postprocessing.ADVANTAGES] + states_entropy)
+            train_batch[Postprocessing.VALUE_TARGETS] = (
+                train_batch[Postprocessing.VALUE_TARGETS] + states_entropy)
+        super().on_learn_on_batch(
+            policy=policy, train_batch=train_batch, result=result, **kwargs)
+
+    def on_train_result(self, *, trainer, result: dict, **kwargs) -> None:
+        RE3UpdateCallbacks._step = result["training_iteration"]
```

Comment:
on_train_result() doesn't necessarily fire after each iteration though ...
I think the best way to do this is to pass trainer._iteration as a parameter to on_learn_on_batch() call.
that way, you don't even need to track it yourself.

maybe you can add a TODO here for me. I am happy to refactor and clean this up once this PR is merged. it feels like a bigger change that we should help you guys.

Reply:
I see. I was not aware of a best way to access trainer._iteration inside `on_learn_on_batch` call. It would be great if refactor it as required.